ouroboros-network/test/Test/Socket.hs

```diff
@@ -69,182 +83,217 @@ prop_socket_demo (TestBlockChainAndUpdates chain updates) =
+-- a read.
+prop_socket_recv_close :: (Int -> Int -> (Int, Int))
+                       -> [Int]
                        -> Property
-prop_socket_recv_close request response = ioProperty $ do
-    b:_ <- getAddrInfo Nothing (Just "127.0.0.1") (Just "6061")
-
-    endMpsVar <- atomically $ newTVar 2
-    resq <- atomically $ newTBQueue 1
-
-    (_, _, server_mp) <- Mxt.setupMiniReqRsp (return ()) endMpsVar request response
-
-    let server_mps Mxt.ChainSync1 = server_mp
-
-    server_h <- startResponderT server_mps b (Just $ rescb resq)
-
-    sd <- socket (addrFamily b) Stream defaultProtocol
-    connect sd (addrAddress b)
-
-    Socket.sendAll sd $ BL.singleton 0xa
-    close sd
-
-    res <- atomically $ readTBQueue resq
-
-    killResponder server_h
-    case res of
-         Just e  ->
-             case fromException e of
-                  Just me -> return $ Mx.errorType me == Mx.MuxBearerClosed
-                  Nothing -> return False
-         Nothing -> return False
-
-  where
-    rescb resq e_m = atomically $ writeTBQueue resq e_m
-
-prop_socket_client_connect_error :: Mxt.DummyPayload
-                                 -> Mxt.DummyPayload
+prop_socket_recv_close f _ = ioProperty $ do
+
+    sv   <- newEmptyTMVarM
+
+    let srvPeer :: Peer (ReqResp.ReqResp Int Int) AsServer ReqResp.StIdle IO ()
+        srvPeer = ReqResp.reqRespServerPeer (reqRespServerMapAccumL sv (\a -> pure . f a) 0)
+        srvPeers Mxt.ReqResp1 = OnlyServer nullTracer ReqResp.codecReqResp srvPeer
+
+    bracket
+      (Socket.socket Socket.AF_INET Socket.Stream Socket.defaultProtocol)
+      Socket.close
+      $ \sd -> do
+        -- bind the socket
+        muxAddress:_ <- Socket.getAddrInfo Nothing (Just "127.0.0.1") (Just "6061")
+        Socket.setSocketOption sd Socket.ReuseAddr 1
+        Socket.bind sd (Socket.addrAddress muxAddress)
+        Socket.listen sd 1
+
+        withAsync
+          (do
+              -- accept a connection and start mux on it
+              bracket
+                (Socket.accept sd)
+                (\(sd',_) -> Socket.close sd')
+                $ \(sd',_) -> do
+                  bearer <- socketAsMuxBearer sd'
+                  Mx.muxBearerSetState bearer Mx.Connected
+                  Mx.muxStart (miniProtocolDescription . srvPeers) bearer
+          )
+          $ \muxAsync -> do
+
+          -- connect to muxAddress
+          sd' <- Socket.socket (Socket.addrFamily muxAddress) Socket.Stream Socket.defaultProtocol
+          Socket.connect sd' (Socket.addrAddress muxAddress)
+
+          Socket.sendAll sd' $ BL.singleton 0xa
+          Socket.close sd'
+
+          res <- waitCatch muxAsync
+          case res of
+              Left e  ->
+                  case fromException e of
+                        Just me -> return $ Mx.errorType me === Mx.MuxBearerClosed
+                        Nothing -> return $ counterexample (show e) False
+              Right _ -> return $ property $ False
+
```

Comment:
It seems like we require a lot of custom code to be able to test that we throw the correct exception which means that we don't really test `Socket.hs` or `Server/Socket`.

Reply:
Indeed we don't, this just tests that `muxStart` throws the right exception, I'll open an issue to rewrite it.